services/versioncache/versioncache.go

```diff
@@ -0,0 +1,284 @@
+		if err != nil {
+			return err
+		}
+		if len(pmmAgents) == 0 {
+			return errors.Errorf("pmmAgent not found for service")
+		}
+		results.AgentID = pmmAgents[0].AgentID
+
+		checkAt := time.Now().UTC().Add(serviceCheckInterval) // TODO
+		if _, err := models.UpdateServiceSoftwareVersions(tx.Querier, servicesVersions[0].ID,
+			models.UpdateServiceSoftwareVersionsParams{CheckAt: &checkAt},
+		); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+func (s *Service) updateVersions() (time.Duration, error) {
+	r, err := s.prepareUpdateVersions()
+	if err != nil {
+		return relaxDuration, err
+	}
+
+	if !r.NeedsUpdate {
+		return r.CheckAfter, nil
+	}
+
+	softwares := []agents.Software{&agents.Mysqld{}, &agents.Xtrabackup{}, &agents.Xbcloud{}, &agents.Qpress{}}
+	versions, err := s.v.GetVersions(r.AgentID, softwares)
+	if err != nil {
+		return relaxDuration, err
+	}
+	if len(versions) != len(softwares) {
+		return relaxDuration, errors.Errorf("slices length mismatch: versions len %d != softwares len %d",
+			len(versions), len(softwares))
+	}
+
+	svs := make([]models.SoftwareVersion, 0, len(softwares))
+	for i, software := range softwares {
+		var softwareName models.SoftwareName
+		switch software := software.(type) {
+		case *agents.Mysqld:
+			softwareName = models.MysqldSoftwareName
+		case *agents.Xtrabackup:
+			softwareName = models.XtrabackupSoftwareName
+		case *agents.Xbcloud:
+			softwareName = models.XbcloudSoftwareName
+		case *agents.Qpress:
+			softwareName = models.QpressSoftwareName
+		default:
+			return relaxDuration, errors.Errorf("invalid software type %T", software)
+		}
+
+		if versions[i].Error != "" {
+			s.l.Warnf("failed to get version of %q software: %s", softwareName, versions[i].Error)
+			continue
+		}
+		if versions[i].Version == "" {
+			continue
+		}
+
+		svs = append(svs, models.SoftwareVersion{
+			Name:    softwareName,
+			Version: versions[i].Version,
+		})
+	}
+
+	if _, err := models.UpdateServiceSoftwareVersions(s.db.Querier, r.ID,
+		models.UpdateServiceSoftwareVersionsParams{Versions: &svs},
+	); err != nil {
+		return relaxDuration, err
+	}
+
+	return relaxDuration, err
+}
+
+// SyncAndUpdate triggers sync and update service software versions.
+func (s *Service) SyncAndUpdate() {
+	select {
+	case s.updateCh <- struct{}{}:
+	default:
+	}
+}
+
+// Run runs software version cache service.
+func (s *Service) Run(ctx context.Context) {
+	defer close(s.updateCh)
+
+	if err := s.syncServices(); err != nil {
+		s.l.Warn(err)
+	}
+
+	syncTicker := time.NewTicker(serviceCheckInterval)
+	defer syncTicker.Stop()
+
+	var checkAfter time.Duration
+	for {
+		select {
+		case <-time.After(checkAfter):
+			ca, err := s.updateVersions()
+			if err != nil {
+				s.l.Warn(err)
+			}
+
+			checkAfter = ca
+		case <-syncTicker.C:
+			if err := s.syncServices(); err != nil {
+				s.l.Warn(err)
+			}
```

Comment:
Mmmm.... why do we need to update services every hour?

Reply:
yeah, it's too frequent. Probably better to check every day?

Reply:
interval increased to one day